File: binary_search/Python/binary_search.py

def binarySearch (arr, l, r, key): 
	"""
    Fuction to do binary search.

    :param 
    	arr: A sorted list of elements,
    	l: indekey of left pointer
    	r: indekey of right pointer
    	key: 
    """
  
    # Check base case 
	if r >= l:
		mid = int(l + (r - l)/2)
  
		# If element is present at the middle itself 
		if arr[mid] == key: 
			return mid 
          
       	# If element is smaller than mid, then it  
        # can only be present in left subarray 
		elif arr[mid] > key: 
			return binarySearch(arr, l, mid-1, key) 
  
        # Else the element can only be present  
        # in right subarray 
		else: 
			return binarySearch(arr, mid+1, r, key) 
	else: 
        # Element is not present in the array 
		return -1

  
def main():
	# Test array 
	arr = [ 2, 3, 4, 10, 40 ] 
	key = 10
  
	# Function call 
	result = binarySearch(arr, 0, len(arr)-1, key) 
	  
	if result != -1: 
	    print("Element is present at index %d" % result)
	else: 
	    print("Element is not present in array")


if __name__ == "__main__":
	main()

File: binary_search/Rust/binary_search.rs
fn binary_search(array: &[i64], length: usize, target: i64) -> i64 {
    let mut left = 0;
    let mut right = length-1;

    while left <= right {
        let middle = (left + right) / 2;
        if array[middle] < target {
            left = middle + 1;
        } else if array[middle] > target {
            right = middle - 1;
        } else {
            return middle as i64;
        }
    }

    return -1;
}

fn main() {
    let array = [1, 4, 16, 21, 109];
    let length = array.len();
    let target = 4;
    
    let index = binary_search(&array, length, target);
    
    println!("Index of {} in array is: {}", target, index);
}
